### ISSUE TITLE
intake-server "can not serialize 'Timestamp' object" when we have datetime parameters

### ISSUE THREAD

**Issue by sliem** [intake/intake#374]
First of all, really cool project!

I'm trying to run an intake-server with an time series data set. It uses parameters of `type: datetime`. The catalog yaml file looks something like this

```yaml
sources:
  timeseriessource:
    driver: sometimeseriesapi
    description: get some time series data
    args:
      label: "{{label}}"
      start: "{{start}}"
      end: "{{end}}"
    parameters:
      label:
        description: data identifier
        type: str
      start:
        description: start time of the time series
        type: datetime
      end:
        description: start time of the time series
        type: datetime
```

It works locally but when using `intake-server` it fails with a `RemoteCatalogError: Failed to fetch metadata.` In the `intake-server` logs we see this traceback
```python
Traceback (most recent call last):
  File "/home/sliem/intake/.venv/lib/python3.7/site-packages/tornado/web.py", line 1697, in _execute
    result = method(*self.path_args, **self.path_kwargs)
  File "/home/sliem/intake/.venv/lib/python3.7/site-packages/intake/cli/server/server.py", line 132, in get
    self.write(msgpack.packb(server_info, use_bin_type=True))
  File "/home/sliem/intake/.venv/lib/python3.7/site-packages/msgpack/__init__.py", line 46, in
packb
    return Packer(**kwargs).pack(o)
  File "msgpack/_packer.pyx", line 282, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 288, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 285, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 232, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 261, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 232, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 261, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 232, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 279, in msgpack._cmsgpack.Packer._pack
TypeError: can not serialize 'Timestamp' object
```

I noticed that `open_catalog` seems to inserts default values if they aren't specified. For `datetime` it is `Timestamp('1970-01-01 00:00:00')` which isn't serialisable by msgpack and the above exception is triggered. 

```python
>>> intake.open_catalog("datetime.yml").items()
[...]
user_parameters: [
  {'name': 'label', 'description': 'data identifier', 'type': 'str', 'default': ''},
  {'name': 'start', 'description': 'start time of the time series', 'type': 'datetime', 'default': Timestamp('1970-01-01 00:00:00')},
  {'name': 'end', 'description': 'start time of the time series', 'type': 'datetime', 'default': Timestamp('1970-01-01 00:00:00')}]
[...]
```

There is an easy workaround by having `start` and `end` being string (ISO data format) or integer (epoch timestamp) and let the `DataSource` parse them but that isn't very nice for the users.

**Comment by martindurant** [comment]
> There is an easy workaround

Would it be complicated to make this addition to the set of msgpack de/serializers?

**Comment by martindurant** [comment]
Some discussion: https://github.com/msgpack/msgpack-python/issues/252 (with further links to open issues, but no agreed solution)

**Comment by martindurant** [comment]
The msgpack-python README gives
```
import datetime
import msgpack

useful_dict = {
    "id": 1,
    "created": datetime.datetime.now(),
}

def decode_datetime(obj):
    if b'__datetime__' in obj:
        obj = datetime.datetime.strptime(obj["as_str"], "%Y%m%dT%H:%M:%S.%f")
    return obj

def encode_datetime(obj):
    if isinstance(obj, datetime.datetime):
        return {'__datetime__': True, 'as_str': obj.strftime("%Y%m%dT%H:%M:%S.%f")}
    return obj
```

**Comment by sliem** [comment]
Changing the `packb` and `unpackb` in server and client to support it would definitely be better approach, yes.

I would however add timezone to the strftime format string. So `"%Y%m%dT%H:%M:%S.%f%z"`.

**Comment by sliem** [comment]
I implemented the `encode_datetime` and `decode_datetime` in https://github.com/sliem/intake/commit/9b3cbca8c2c146499f417dcdaeb595605d7f2507

I won't claim it is tested but it does seem to work.

**Comment by martindurant** [comment]
Would you like to submit that as a PR? Adding a tests specifically for datetime should not be hard, there are other similar tests of remote data sources already.

**Comment by sliem** [comment]
Will do when I got a moment!

**Comment by danielballan** [comment]
I think this can be marked as closed by #376.
